{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "db694681",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install PhenoGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fdc71c46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/bokai/miniconda3/envs/muse/lib/python3.6/site-packages/tensorflow/python/compat/v2_compat.py:101: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "non-resource variables are not supported in the long term\n"
     ]
    }
   ],
   "source": [
    "import muse_sc as muse\n",
    "import phenograph\n",
    "from sklearn.decomposition import PCA\n",
    "import numpy as np\n",
    "from sklearn.metrics.cluster import adjusted_rand_score\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.manifold import TSNE\n",
    "import tensorflow as tf\n",
    "tf.get_logger().setLevel('ERROR')\n",
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "da44837f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/bokai/miniconda3/envs/muse/bin/python'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.executable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "56fadaf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "###### process protein modality\n",
    "\n",
    "data1 = pd.read_csv(\"../data/Xton_sub_norm.csv\")\n",
    "dataa = data1\n",
    "\n",
    "## change to scaled version\n",
    "from scipy.stats import zscore\n",
    "dataa = dataa.apply(zscore)\n",
    "dataa = dataa\n",
    "\n",
    "####### process imaging modality\n",
    "\n",
    "datab_full = np.load('../data/muse_images/Inception_img_feature_xton.npy')\n",
    "datab_full = np.array(datab_full).squeeze(1)\n",
    "\n",
    "###### this is pca on feature mod and image mod\n",
    "\n",
    "latent_dim = 20\n",
    "view_a_feature = PCA(n_components=latent_dim).fit_transform(dataa)\n",
    "### this is pca on img mod\n",
    "latent_dim = 100\n",
    "view_b_feature = PCA(n_components=latent_dim).fit_transform(datab_full)\n",
    "### then default clustering\n",
    "\n",
    "# already finished running\n",
    "view_a_label, _, _ = phenograph.cluster(view_a_feature, n_jobs = 16)\n",
    "view_b_label, _, _ = phenograph.cluster(view_b_feature, n_jobs = 16)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99de16ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('../data/musexton_a_label.npy', view_a_label)\n",
    "np.save('../data/musexton_b_label.npy', view_b_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f20728a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "view_a_label = np.load('../data/musexton_a_label.npy')\n",
    "view_b_label = np.load('../data/musexton_b_label.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dfa5d3b",
   "metadata": {},
   "source": [
    "### change original muse function phenograph cpu allocation too many issue "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "376f5a74",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from muse_sc.muse_architecture import structured_embedding\n",
    "from scipy.spatial.distance import pdist\n",
    "import phenograph\n",
    "import tensorflow.compat.v1 as tf\n",
    "\n",
    "tf.disable_v2_behavior()\n",
    "\n",
    "def muse_fit_predict(data_x,\n",
    "                     data_y,\n",
    "                     label_x,\n",
    "                     label_y,\n",
    "                     latent_dim=100,\n",
    "                     n_epochs=500,\n",
    "                     lambda_regul=5,\n",
    "                     lambda_super=5):\n",
    "    \"\"\"\n",
    "        MUSE model fitting and predicting:\n",
    "          This function is used to train the MUSE model on multi-modality data\n",
    "\n",
    "        Parameters:\n",
    "          data_x:       input for transcript modality; matrix of  n * p, where n = number of cells, p = number of genes.\n",
    "          data_y:       input for morphological modality; matrix of n * q, where n = number of cells, q is the feature dimension.\n",
    "          label_x:      initial reference cluster label for transcriptional modality.\n",
    "          label_y:      inital reference cluster label for morphological modality.\n",
    "          latent_dim:   feature dimension of joint latent representation.\n",
    "          n_epochs:     maximal epoch used in training.\n",
    "          lambda_regul: weight for regularization term in the loss function.\n",
    "          lambda_super: weight for supervised learning loss in the loss function.\n",
    "\n",
    "        Output:\n",
    "          latent:       joint latent representation learned by MUSE.\n",
    "          reconstruct_x:reconstructed feature matrix corresponding to input data_x.\n",
    "          reconstruct_y:reconstructed feature matrix corresponding to input data_y.\n",
    "          latent_x:     modality-specific latent representation corresponding to data_x.\n",
    "          latent_y:     modality-specific latent representation corresponding to data_y.\n",
    "\n",
    "        Feng Bao @ Altschuler & Wu Lab @ UCSF 2022.\n",
    "        Software provided as is under MIT License.\n",
    "    \"\"\"\n",
    "\n",
    "    \"\"\" initial parameter setting \"\"\"\n",
    "    # parameter setting for neural network\n",
    "    n_hidden = 128  # number of hidden node in neural network\n",
    "    learn_rate = 1e-4  # learning rate in the optimization\n",
    "    batch_size = 64  # number of cells in the training batch\n",
    "    n_epochs_init = 200  # number of training epoch in model initialization\n",
    "    print_epochs = 50  # epoch interval to display the current training loss\n",
    "    cluster_update_epoch = 200  # epoch interval to update modality-specific clusters\n",
    "\n",
    "    # read data-specific parameters from inputs\n",
    "    feature_dim_x = data_x.shape[1]\n",
    "    feature_dim_y = data_y.shape[1]\n",
    "    n_sample = data_x.shape[0]\n",
    "\n",
    "    # GPU configuration\n",
    "    # config = tf.ConfigProto()\n",
    "    # config.gpu_options.allow_growth = True\n",
    "\n",
    "    \"\"\" construct computation graph using TensorFlow \"\"\"\n",
    "    tf.reset_default_graph()\n",
    "\n",
    "    # raw data from two modalities\n",
    "    x = tf.placeholder(tf.float32, shape=[None, feature_dim_x], name='input_x')\n",
    "    y = tf.placeholder(tf.float32, shape=[None, feature_dim_y], name='input_y')\n",
    "\n",
    "    # labels inputted for references\n",
    "    ref_label_x = tf.placeholder(tf.float32, shape=[None], name='ref_label_x')\n",
    "    ref_label_y = tf.placeholder(tf.float32, shape=[None], name='ref_label_y')\n",
    "\n",
    "    # hyperparameter in triplet loss\n",
    "    triplet_lambda = tf.placeholder(tf.float32, name='triplet_lambda')\n",
    "    triplet_margin = tf.placeholder(tf.float32, name='triplet_margin')\n",
    "\n",
    "    # network architecture\n",
    "    z, x_hat, y_hat, encode_x, encode_y, loss, \\\n",
    "    reconstruction_error, weight_penalty, \\\n",
    "    trip_loss_x, trip_loss_y = structured_embedding(x,\n",
    "                                                    y,\n",
    "                                                    ref_label_x,\n",
    "                                                    ref_label_y,\n",
    "                                                    latent_dim,\n",
    "                                                    triplet_margin,\n",
    "                                                    n_hidden,\n",
    "                                                    lambda_regul,\n",
    "                                                    triplet_lambda)\n",
    "    # optimization operator\n",
    "    train_op = tf.train.AdamOptimizer(learn_rate).minimize(loss)\n",
    "    print('++++++++++ MUSE for multi-modality single-cell analysis ++++++++++')\n",
    "    \"\"\" MUSE optimization \"\"\"\n",
    "    total_batch = int(n_sample / batch_size)\n",
    "\n",
    "    with tf.Session() as sess:\n",
    "\n",
    "        \"\"\" initialization of autoencoder architecture for MUSE \"\"\"\n",
    "        print('MUSE initialization')\n",
    "        # global parameter initialization\n",
    "        sess.run(tf.global_variables_initializer(), feed_dict={triplet_lambda: 0,\n",
    "                                                               triplet_margin: 0})\n",
    "\n",
    "        for epoch in range(n_epochs_init):\n",
    "            # randomly permute samples\n",
    "            random_idx = np.random.permutation(n_sample)\n",
    "            data_train_x = data_x[random_idx, :]\n",
    "            data_train_y = data_y[random_idx, :]\n",
    "\n",
    "            for i in range(total_batch):\n",
    "                # input data batches\n",
    "                offset = (i * batch_size) % (n_sample)\n",
    "                batch_x_input = data_train_x[offset:(offset + batch_size), :]\n",
    "                batch_y_input = data_train_y[offset:(offset + batch_size), :]\n",
    "\n",
    "                # initialize parameters without self-supervised loss (triplet_lambda=0)\n",
    "                sess.run(train_op,\n",
    "                         feed_dict={x: batch_x_input,\n",
    "                                    y: batch_y_input,\n",
    "                                    ref_label_x: np.zeros(batch_x_input.shape[0]),\n",
    "                                    ref_label_y: np.zeros(batch_y_input.shape[0]),\n",
    "                                    triplet_lambda: 0,\n",
    "                                    triplet_margin: 0})\n",
    "\n",
    "            # calculate and print loss terms for current epoch\n",
    "            if epoch % print_epochs == 0:\n",
    "                L_total, L_reconstruction, L_weight = \\\n",
    "                    sess.run((loss, reconstruction_error, weight_penalty),\n",
    "                             feed_dict={x: data_train_x,\n",
    "                                        y: data_train_y,\n",
    "                                        ref_label_x: np.zeros(data_train_x.shape[0]),  # no use as triplet_lambda=0\n",
    "                                        ref_label_y: np.zeros(data_train_y.shape[0]),  # no use as triplet_lambda=0\n",
    "                                        triplet_lambda: 0,\n",
    "                                        triplet_margin: 0})\n",
    "\n",
    "                print(\n",
    "                    \"epoch: %d, \\t total loss: %03.5f,\\t reconstruction loss: %03.5f,\\t sparse penalty: %03.5f\"\n",
    "                    % (epoch, L_total, L_reconstruction, L_weight))\n",
    "\n",
    "        # estimate the margin for the triplet loss\n",
    "        latent, reconstruct_x, reconstruct_y = \\\n",
    "            sess.run((z, x_hat, y_hat),\n",
    "                     feed_dict={x: data_x,\n",
    "                                y: data_y,\n",
    "                                ref_label_x: np.zeros(data_x.shape[0]),\n",
    "                                ref_label_y: np.zeros(data_y.shape[0]),\n",
    "                                triplet_lambda: 0,\n",
    "                                triplet_margin: 0})\n",
    "        latent_pd_matrix = pdist(latent, 'euclidean')\n",
    "        latent_pd_sort = np.sort(latent_pd_matrix)\n",
    "        select_top_n = np.int(latent_pd_sort.size * 0.2)\n",
    "        margin_estimate = np.median(latent_pd_sort[-select_top_n:]) - np.median(latent_pd_sort[:select_top_n])\n",
    "\n",
    "        # refine MUSE parameters with reference labels and triplet losses\n",
    "        for epoch in range(n_epochs_init):\n",
    "            # randomly permute samples\n",
    "            random_idx = np.random.permutation(n_sample)\n",
    "            data_train_x = data_x[random_idx, :]\n",
    "            data_train_y = data_y[random_idx, :]\n",
    "            label_train_x = label_x[random_idx]\n",
    "            label_train_y = label_y[random_idx]\n",
    "\n",
    "            for i in range(total_batch):\n",
    "                # data batches\n",
    "                offset = (i * batch_size) % (n_sample)\n",
    "                batch_x_input = data_train_x[offset:(offset + batch_size), :]\n",
    "                batch_y_input = data_train_y[offset:(offset + batch_size), :]\n",
    "                label_x_input = label_train_x[offset:(offset + batch_size)]\n",
    "                label_y_input = label_train_y[offset:(offset + batch_size)]\n",
    "\n",
    "                # refine parameters\n",
    "                sess.run(train_op,\n",
    "                         feed_dict={x: batch_x_input,\n",
    "                                    y: batch_y_input,\n",
    "                                    ref_label_x: label_x_input,\n",
    "                                    ref_label_y: label_y_input,\n",
    "                                    triplet_lambda: lambda_super,\n",
    "                                    triplet_margin: margin_estimate})\n",
    "\n",
    "            # calculate loss on all input data for current epoch\n",
    "            if epoch % print_epochs == 0:\n",
    "                L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y = \\\n",
    "                    sess.run((loss, reconstruction_error, weight_penalty, trip_loss_x, trip_loss_y),\n",
    "                             feed_dict={x: data_train_x,\n",
    "                                        y: data_train_y,\n",
    "                                        ref_label_x: label_train_x,\n",
    "                                        ref_label_y: label_train_y,\n",
    "                                        triplet_lambda: lambda_super,\n",
    "                                        triplet_margin: margin_estimate})\n",
    "\n",
    "                print(\n",
    "                    \"epoch: %d, \\t total loss: %03.5f,\\t reconstruction loss: %03.5f,\\t sparse penalty: %03.5f,\\t x triplet: %03.5f,\\t y triplet: %03.5f\"\n",
    "                    % (epoch, L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y))\n",
    "\n",
    "        # update cluster labels based modality-specific latents\n",
    "        latent_x, latent_y = \\\n",
    "            sess.run((encode_x, encode_y),\n",
    "                     feed_dict={x: data_x,\n",
    "                                y: data_y,\n",
    "                                ref_label_x: label_x,\n",
    "                                ref_label_y: label_y,\n",
    "                                triplet_lambda: lambda_super,\n",
    "                                triplet_margin: margin_estimate})\n",
    "\n",
    "        # update cluster labels using PhenoGraph\n",
    "        label_x_update, _, _ = phenograph.cluster(latent_x, n_jobs= 1)\n",
    "        label_y_update, _, _ = phenograph.cluster(latent_y, n_jobs= 1)\n",
    "        print('Finish initialization of MUSE')\n",
    "\n",
    "        ''' Training of MUSE '''\n",
    "        for epoch in range(n_epochs):\n",
    "            # randomly permute samples\n",
    "            random_idx = np.random.permutation(n_sample)\n",
    "            data_train_x = data_x[random_idx, :]\n",
    "            data_train_y = data_y[random_idx, :]\n",
    "            label_train_x = label_x_update[random_idx]\n",
    "            label_train_y = label_y_update[random_idx]\n",
    "\n",
    "            # loop over all batches\n",
    "            for i in range(total_batch):\n",
    "                # batch data\n",
    "                offset = (i * batch_size) % (n_sample)\n",
    "                batch_x_input = data_train_x[offset:(offset + batch_size), :]\n",
    "                batch_y_input = data_train_y[offset:(offset + batch_size), :]\n",
    "                batch_label_x_input = label_train_x[offset:(offset + batch_size)]\n",
    "                batch_label_y_input = label_train_y[offset:(offset + batch_size)]\n",
    "\n",
    "                sess.run(train_op,\n",
    "                         feed_dict={x: batch_x_input,\n",
    "                                    y: batch_y_input,\n",
    "                                    ref_label_x: batch_label_x_input,\n",
    "                                    ref_label_y: batch_label_y_input,\n",
    "                                    triplet_lambda: lambda_super,\n",
    "                                    triplet_margin: margin_estimate})\n",
    "\n",
    "            # calculate and print losses on whole training dataset\n",
    "            if epoch % print_epochs == 0:\n",
    "                L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y = \\\n",
    "                    sess.run((loss, reconstruction_error, weight_penalty, trip_loss_x, trip_loss_y),\n",
    "                             feed_dict={x: data_train_x,\n",
    "                                        y: data_train_y,\n",
    "                                        ref_label_x: label_train_x,\n",
    "                                        ref_label_y: label_train_y,\n",
    "                                        triplet_lambda: lambda_super,\n",
    "                                        triplet_margin: margin_estimate})\n",
    "                # print cost every epoch\n",
    "                print(\n",
    "                    \"epoch: %d, \\t total loss: %03.5f,\\t reconstruction loss: %03.5f,\\t sparse penalty: %03.5f,\\t x triplet loss: %03.5f,\\t y triplet loss: %03.5f\"\n",
    "                    % (epoch, L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y))\n",
    "\n",
    "            # update cluster labels based on new modality-specific latent representations\n",
    "            if epoch % cluster_update_epoch == 0:\n",
    "                latent_x, latent_y = \\\n",
    "                    sess.run((encode_x, encode_y),\n",
    "                             feed_dict={x: data_x,\n",
    "                                        y: data_y,\n",
    "                                        ref_label_x: label_x,\n",
    "                                        ref_label_y: label_y,\n",
    "                                        triplet_lambda: lambda_super,\n",
    "                                        triplet_margin: margin_estimate})\n",
    "\n",
    "                # use PhenoGraph to obtain cluster label\n",
    "                label_x_update, _, _ = phenograph.cluster(latent_x, n_jobs= 1)\n",
    "                label_y_update, _, _ = phenograph.cluster(latent_y, n_jobs= 1)\n",
    "\n",
    "        \"\"\" MUSE output \"\"\"\n",
    "        latent, reconstruct_x, reconstruct_y, latent_x, latent_y = \\\n",
    "            sess.run((z, x_hat, y_hat, encode_x, encode_y),\n",
    "                     feed_dict={x: data_x,\n",
    "                                y: data_y,\n",
    "                                ref_label_x: label_x,  # no effects to representations\n",
    "                                ref_label_y: label_y,  # no effects to representations\n",
    "                                triplet_lambda: lambda_super,\n",
    "                                triplet_margin: margin_estimate})\n",
    "\n",
    "        print('++++++++++ MUSE completed ++++++++++')\n",
    "\n",
    "    return latent, reconstruct_x, reconstruct_y, latent_x, latent_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2deeb76a",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1770.50745,\t reconstruction loss: 1479.88721,\t sparse penalty: 58.12405\n",
      "epoch: 50, \t total loss: 899.01154,\t reconstruction loss: 868.85266,\t sparse penalty: 6.03177\n",
      "epoch: 100, \t total loss: 845.44275,\t reconstruction loss: 844.69458,\t sparse penalty: 0.14963\n",
      "epoch: 150, \t total loss: 825.85522,\t reconstruction loss: 825.39850,\t sparse penalty: 0.09135\n",
      "epoch: 0, \t total loss: 1433.80237,\t reconstruction loss: 1190.69580,\t sparse penalty: 0.22681,\t x triplet: 24.45541,\t y triplet: 23.93909\n",
      "epoch: 50, \t total loss: 1096.45801,\t reconstruction loss: 871.16827,\t sparse penalty: 0.40197,\t x triplet: 22.41992,\t y triplet: 22.23607\n",
      "epoch: 100, \t total loss: 1084.74194,\t reconstruction loss: 861.36346,\t sparse penalty: 0.34760,\t x triplet: 22.23248,\t y triplet: 22.09562\n",
      "epoch: 150, \t total loss: 1079.02161,\t reconstruction loss: 856.43817,\t sparse penalty: 0.30979,\t x triplet: 22.16194,\t y triplet: 22.04496\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.503740310668945 seconds\n",
      "Jaccard graph constructed in 16.595695734024048 seconds\n",
      "Wrote graph to binary file in 1.9311389923095703 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.646822\n",
      "After 2 runs, maximum modularity is Q = 0.649911\n",
      "After 3 runs, maximum modularity is Q = 0.65476\n",
      "After 4 runs, maximum modularity is Q = 0.656119\n",
      "Louvain completed 24 runs in 17.802303314208984 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 57.37076497077942 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.209380865097046 seconds\n",
      "Jaccard graph constructed in 16.5775887966156 seconds\n",
      "Wrote graph to binary file in 2.3016626834869385 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.815912\n",
      "After 5 runs, maximum modularity is Q = 0.817293\n",
      "Louvain completed 25 runs in 20.18136763572693 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 60.467308044433594 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 1077.68542,\t reconstruction loss: 855.48126,\t sparse penalty: 0.28511,\t x triplet loss: 22.09658,\t y triplet loss: 22.05914\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 11.088051557540894 seconds\n",
      "Jaccard graph constructed in 18.049811601638794 seconds\n",
      "Wrote graph to binary file in 2.8275325298309326 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.670159\n",
      "Louvain completed 21 runs in 12.350480318069458 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 55.292038679122925 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 11.328692197799683 seconds\n",
      "Jaccard graph constructed in 19.061222314834595 seconds\n",
      "Wrote graph to binary file in 2.303037166595459 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.825637\n",
      "After 4 runs, maximum modularity is Q = 0.827048\n",
      "After 19 runs, maximum modularity is Q = 0.82822\n",
      "Louvain completed 39 runs in 31.136894702911377 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 75.63355731964111 seconds\n",
      "epoch: 50, \t total loss: 1071.74866,\t reconstruction loss: 850.27234,\t sparse penalty: 0.27556,\t x triplet loss: 22.02484,\t y triplet loss: 21.99486\n",
      "epoch: 100, \t total loss: 1071.15479,\t reconstruction loss: 849.99207,\t sparse penalty: 0.25493,\t x triplet loss: 22.00032,\t y triplet loss: 21.97729\n",
      "epoch: 150, \t total loss: 1069.46960,\t reconstruction loss: 848.57446,\t sparse penalty: 0.23485,\t x triplet loss: 21.98296,\t y triplet loss: 21.96123\n",
      "epoch: 200, \t total loss: 1068.75854,\t reconstruction loss: 848.08246,\t sparse penalty: 0.21761,\t x triplet loss: 21.96879,\t y triplet loss: 21.94883\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.91094970703125 seconds\n",
      "Jaccard graph constructed in 18.408056497573853 seconds\n",
      "Wrote graph to binary file in 2.3275930881500244 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.713462\n",
      "After 3 runs, maximum modularity is Q = 0.715741\n",
      "After 4 runs, maximum modularity is Q = 0.7174\n",
      "Louvain completed 24 runs in 20.00977349281311 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 64.37450671195984 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 12.60456109046936 seconds\n",
      "Jaccard graph constructed in 18.317787408828735 seconds\n",
      "Wrote graph to binary file in 2.11948299407959 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.846412\n",
      "Louvain completed 21 runs in 16.13271403312683 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 62.2080340385437 seconds\n",
      "epoch: 250, \t total loss: 1060.96997,\t reconstruction loss: 840.50238,\t sparse penalty: 0.20235,\t x triplet loss: 21.95473,\t y triplet loss: 21.93644\n",
      "epoch: 300, \t total loss: 1058.73425,\t reconstruction loss: 838.45038,\t sparse penalty: 0.19142,\t x triplet loss: 21.94027,\t y triplet loss: 21.92509\n",
      "epoch: 350, \t total loss: 1057.80408,\t reconstruction loss: 837.65778,\t sparse penalty: 0.18261,\t x triplet loss: 21.93149,\t y triplet loss: 21.91515\n",
      "epoch: 400, \t total loss: 1056.61572,\t reconstruction loss: 836.59698,\t sparse penalty: 0.17622,\t x triplet loss: 21.92040,\t y triplet loss: 21.90713\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.583301782608032 seconds\n",
      "Jaccard graph constructed in 20.748183488845825 seconds\n",
      "Wrote graph to binary file in 2.241166830062866 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.719979\n",
      "Louvain completed 21 runs in 14.028547048568726 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 60.83184599876404 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 13.196720123291016 seconds\n",
      "Jaccard graph constructed in 21.570637941360474 seconds\n",
      "Wrote graph to binary file in 1.97825288772583 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.848922\n",
      "After 2 runs, maximum modularity is Q = 0.850734\n",
      "After 7 runs, maximum modularity is Q = 0.852521\n",
      "Louvain completed 27 runs in 21.726207494735718 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 72.19646835327148 seconds\n",
      "epoch: 450, \t total loss: 1049.38318,\t reconstruction loss: 829.40851,\t sparse penalty: 0.17637,\t x triplet loss: 21.92242,\t y triplet loss: 21.89613\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1771.55981,\t reconstruction loss: 1478.25879,\t sparse penalty: 58.66019\n",
      "epoch: 50, \t total loss: 896.47998,\t reconstruction loss: 870.28650,\t sparse penalty: 5.23869\n",
      "epoch: 100, \t total loss: 858.38361,\t reconstruction loss: 857.89764,\t sparse penalty: 0.09719\n",
      "epoch: 150, \t total loss: 840.08258,\t reconstruction loss: 839.71881,\t sparse penalty: 0.07275\n",
      "epoch: 0, \t total loss: 1252.50000,\t reconstruction loss: 1007.63818,\t sparse penalty: 0.28436,\t x triplet: 24.57067,\t y triplet: 24.11732\n",
      "epoch: 50, \t total loss: 1095.01379,\t reconstruction loss: 875.18903,\t sparse penalty: 0.38368,\t x triplet: 21.86612,\t y triplet: 21.71515\n",
      "epoch: 100, \t total loss: 1089.78369,\t reconstruction loss: 871.74298,\t sparse penalty: 0.30077,\t x triplet: 21.71786,\t y triplet: 21.58952\n",
      "epoch: 150, \t total loss: 1081.41919,\t reconstruction loss: 864.21106,\t sparse penalty: 0.25159,\t x triplet: 21.64701,\t y triplet: 21.54301\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 8.451819658279419 seconds\n",
      "Jaccard graph constructed in 24.487616539001465 seconds\n",
      "Wrote graph to binary file in 2.45662522315979 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.719819\n",
      "After 6 runs, maximum modularity is Q = 0.721402\n",
      "Louvain completed 26 runs in 19.43755316734314 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 70.27713894844055 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.338266134262085 seconds\n",
      "Jaccard graph constructed in 23.52399730682373 seconds\n",
      "Wrote graph to binary file in 2.0250308513641357 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.799897\n",
      "Louvain completed 21 runs in 20.454119205474854 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 72.18195962905884 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 1075.97290,\t reconstruction loss: 859.40356,\t sparse penalty: 0.23555,\t x triplet loss: 21.55244,\t y triplet loss: 21.52589\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 7.604003667831421 seconds\n",
      "Jaccard graph constructed in 24.28178334236145 seconds\n",
      "Wrote graph to binary file in 2.748838186264038 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.725423\n",
      "Louvain completed 21 runs in 15.353637218475342 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 66.92222857475281 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 17.97186803817749 seconds\n",
      "Jaccard graph constructed in 24.614516973495483 seconds\n",
      "Wrote graph to binary file in 2.2640700340270996 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.817054\n",
      "After 4 runs, maximum modularity is Q = 0.818934\n",
      "After 17 runs, maximum modularity is Q = 0.820848\n",
      "Louvain completed 37 runs in 33.65195631980896 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 96.47274231910706 seconds\n",
      "epoch: 50, \t total loss: 1071.29639,\t reconstruction loss: 855.39636,\t sparse penalty: 0.22194,\t x triplet loss: 21.49161,\t y triplet loss: 21.46646\n",
      "epoch: 100, \t total loss: 1066.97205,\t reconstruction loss: 851.37421,\t sparse penalty: 0.20517,\t x triplet loss: 21.46949,\t y triplet loss: 21.44492\n",
      "epoch: 150, \t total loss: 1064.39893,\t reconstruction loss: 849.01941,\t sparse penalty: 0.19133,\t x triplet loss: 21.45547,\t y triplet loss: 21.42909\n",
      "epoch: 200, \t total loss: 1063.92041,\t reconstruction loss: 848.75134,\t sparse penalty: 0.18074,\t x triplet loss: 21.43740,\t y triplet loss: 21.41566\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 8.813868522644043 seconds\n",
      "Jaccard graph constructed in 26.753327131271362 seconds\n",
      "Wrote graph to binary file in 2.8194427490234375 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.717369\n",
      "After 4 runs, maximum modularity is Q = 0.720431\n",
      "After 18 runs, maximum modularity is Q = 0.721582\n",
      "Louvain completed 38 runs in 29.959517240524292 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 85.69432353973389 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 30.27204918861389 seconds\n",
      "Jaccard graph constructed in 26.136965036392212 seconds\n",
      "Wrote graph to binary file in 2.171732187271118 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.848149\n",
      "After 6 runs, maximum modularity is Q = 0.850092\n",
      "Louvain completed 26 runs in 23.444765090942383 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 99.24226331710815 seconds\n",
      "epoch: 250, \t total loss: 1060.72986,\t reconstruction loss: 845.73419,\t sparse penalty: 0.17003,\t x triplet loss: 21.42159,\t y triplet loss: 21.40752\n",
      "epoch: 300, \t total loss: 1059.02759,\t reconstruction loss: 844.14575,\t sparse penalty: 0.16286,\t x triplet loss: 21.41245,\t y triplet loss: 21.40104\n",
      "epoch: 350, \t total loss: 1057.65601,\t reconstruction loss: 842.88281,\t sparse penalty: 0.15680,\t x triplet loss: 21.40384,\t y triplet loss: 21.39399\n",
      "epoch: 400, \t total loss: 1056.00903,\t reconstruction loss: 841.34259,\t sparse penalty: 0.15058,\t x triplet loss: 21.39585,\t y triplet loss: 21.38686\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 9.444189548492432 seconds\n",
      "Jaccard graph constructed in 26.638883113861084 seconds\n",
      "Wrote graph to binary file in 2.795877695083618 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.710751\n",
      "After 3 runs, maximum modularity is Q = 0.713442\n",
      "After 15 runs, maximum modularity is Q = 0.715123\n",
      "Louvain completed 35 runs in 31.151583194732666 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 88.12736749649048 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 36.04196333885193 seconds\n",
      "Jaccard graph constructed in 26.264786958694458 seconds\n",
      "Wrote graph to binary file in 1.8920185565948486 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.856119\n",
      "After 3 runs, maximum modularity is Q = 0.858598\n",
      "After 23 runs, maximum modularity is Q = 0.861071\n",
      "Louvain completed 43 runs in 34.397944927215576 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 116.78798341751099 seconds\n",
      "epoch: 450, \t total loss: 1058.70935,\t reconstruction loss: 844.14734,\t sparse penalty: 0.14923,\t x triplet loss: 21.38779,\t y triplet loss: 21.37538\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1765.93054,\t reconstruction loss: 1478.57776,\t sparse penalty: 57.47056\n",
      "epoch: 50, \t total loss: 902.46210,\t reconstruction loss: 876.89618,\t sparse penalty: 5.11319\n",
      "epoch: 100, \t total loss: 856.98285,\t reconstruction loss: 856.52795,\t sparse penalty: 0.09098\n",
      "epoch: 150, \t total loss: 842.18365,\t reconstruction loss: 841.84998,\t sparse penalty: 0.06674\n",
      "epoch: 0, \t total loss: 1209.53503,\t reconstruction loss: 975.88202,\t sparse penalty: 0.28158,\t x triplet: 23.45671,\t y triplet: 22.99231\n",
      "epoch: 50, \t total loss: 1072.47314,\t reconstruction loss: 862.28766,\t sparse penalty: 0.40695,\t x triplet: 20.87855,\t y triplet: 20.75158\n",
      "epoch: 100, \t total loss: 1056.62378,\t reconstruction loss: 847.96472,\t sparse penalty: 0.33443,\t x triplet: 20.74901,\t y triplet: 20.64837\n",
      "epoch: 150, \t total loss: 1050.37170,\t reconstruction loss: 842.68127,\t sparse penalty: 0.28878,\t x triplet: 20.66845,\t y triplet: 20.58085\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 19.902579307556152 seconds\n",
      "Jaccard graph constructed in 28.150225400924683 seconds\n",
      "Wrote graph to binary file in 3.047402858734131 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.662162\n",
      "After 18 runs, maximum modularity is Q = 0.664167\n",
      "Louvain completed 38 runs in 28.588836669921875 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 100.06053876876831 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 78.4608371257782 seconds\n",
      "Jaccard graph constructed in 28.04399037361145 seconds\n",
      "Wrote graph to binary file in 1.8944811820983887 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.70795\n",
      "After 3 runs, maximum modularity is Q = 0.713456\n",
      "After 14 runs, maximum modularity is Q = 0.715243\n",
      "After 28 runs, maximum modularity is Q = 0.717914\n",
      "Louvain completed 48 runs in 52.234323501586914 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 181.04352688789368 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 1046.77051,\t reconstruction loss: 839.77155,\t sparse penalty: 0.26506,\t x triplet loss: 20.61080,\t y triplet loss: 20.52393\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 9.762234687805176 seconds\n",
      "Jaccard graph constructed in 29.314729928970337 seconds\n",
      "Wrote graph to binary file in 2.6113460063934326 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.628783\n",
      "After 2 runs, maximum modularity is Q = 0.635857\n",
      "After 4 runs, maximum modularity is Q = 0.638497\n",
      "Louvain completed 24 runs in 22.23499870300293 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 87.4586591720581 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 13.300530433654785 seconds\n",
      "Jaccard graph constructed in 29.19986057281494 seconds\n",
      "Wrote graph to binary file in 2.0220489501953125 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.71536\n",
      "Louvain completed 21 runs in 21.472322702407837 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 87.32476043701172 seconds\n",
      "epoch: 50, \t total loss: 1041.10632,\t reconstruction loss: 834.78418,\t sparse penalty: 0.25120,\t x triplet loss: 20.54611,\t y triplet loss: 20.46711\n",
      "epoch: 100, \t total loss: 1037.10168,\t reconstruction loss: 831.14948,\t sparse penalty: 0.23054,\t x triplet loss: 20.51517,\t y triplet loss: 20.44472\n",
      "epoch: 150, \t total loss: 1035.18079,\t reconstruction loss: 829.53949,\t sparse penalty: 0.21494,\t x triplet loss: 20.48808,\t y triplet loss: 20.42524\n",
      "epoch: 200, \t total loss: 1031.62024,\t reconstruction loss: 826.20721,\t sparse penalty: 0.20189,\t x triplet loss: 20.47058,\t y triplet loss: 20.41014\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.597092628479004 seconds\n",
      "Jaccard graph constructed in 31.228705167770386 seconds\n",
      "Wrote graph to binary file in 2.694842576980591 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.676176\n",
      "After 5 runs, maximum modularity is Q = 0.678275\n",
      "Louvain completed 25 runs in 22.65269112586975 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 89.4473602771759 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 93.10065650939941 seconds\n",
      "Jaccard graph constructed in 31.04967951774597 seconds\n",
      "Wrote graph to binary file in 2.5378146171569824 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.710343\n",
      "After 2 runs, maximum modularity is Q = 0.714748\n",
      "After 4 runs, maximum modularity is Q = 0.717612\n",
      "After 6 runs, maximum modularity is Q = 0.720219\n",
      "Louvain completed 26 runs in 30.71491026878357 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 179.27980875968933 seconds\n",
      "epoch: 250, \t total loss: 1030.32556,\t reconstruction loss: 825.17432,\t sparse penalty: 0.19700,\t x triplet loss: 20.43772,\t y triplet loss: 20.39554\n",
      "epoch: 300, \t total loss: 1027.66821,\t reconstruction loss: 822.67664,\t sparse penalty: 0.18772,\t x triplet loss: 20.42642,\t y triplet loss: 20.38419\n",
      "epoch: 350, \t total loss: 1028.13586,\t reconstruction loss: 823.28723,\t sparse penalty: 0.17951,\t x triplet loss: 20.41529,\t y triplet loss: 20.37493\n",
      "epoch: 400, \t total loss: 1025.13745,\t reconstruction loss: 820.41614,\t sparse penalty: 0.17204,\t x triplet loss: 20.40529,\t y triplet loss: 20.36692\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 9.800575494766235 seconds\n",
      "Jaccard graph constructed in 32.78770923614502 seconds\n",
      "Wrote graph to binary file in 2.558650255203247 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.681635\n",
      "After 15 runs, maximum modularity is Q = 0.682829\n",
      "After 33 runs, maximum modularity is Q = 0.68476\n",
      "Louvain completed 53 runs in 46.22388792037964 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 115.23205590248108 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 43.877689599990845 seconds\n",
      "Jaccard graph constructed in 30.507280588150024 seconds\n",
      "Wrote graph to binary file in 1.9943716526031494 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.713868\n",
      "After 2 runs, maximum modularity is Q = 0.716841\n",
      "After 6 runs, maximum modularity is Q = 0.72185\n",
      "Louvain completed 26 runs in 30.34405779838562 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 130.51917362213135 seconds\n",
      "epoch: 450, \t total loss: 1024.06775,\t reconstruction loss: 819.46552,\t sparse penalty: 0.17105,\t x triplet loss: 20.39244,\t y triplet loss: 20.35695\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1768.46472,\t reconstruction loss: 1481.43616,\t sparse penalty: 57.40571\n",
      "epoch: 50, \t total loss: 901.26361,\t reconstruction loss: 870.62238,\t sparse penalty: 6.12825\n",
      "epoch: 100, \t total loss: 851.40894,\t reconstruction loss: 850.89252,\t sparse penalty: 0.10328\n",
      "epoch: 150, \t total loss: 833.30487,\t reconstruction loss: 832.92914,\t sparse penalty: 0.07514\n",
      "epoch: 0, \t total loss: 1244.32910,\t reconstruction loss: 1001.44916,\t sparse penalty: 0.27290,\t x triplet: 24.41815,\t y triplet: 23.88494\n",
      "epoch: 50, \t total loss: 1082.79407,\t reconstruction loss: 865.85663,\t sparse penalty: 0.39263,\t x triplet: 21.54921,\t y triplet: 21.44566\n",
      "epoch: 100, \t total loss: 1077.08667,\t reconstruction loss: 861.63232,\t sparse penalty: 0.33417,\t x triplet: 21.43315,\t y triplet: 21.32355\n",
      "epoch: 150, \t total loss: 1068.18591,\t reconstruction loss: 853.29523,\t sparse penalty: 0.29834,\t x triplet: 21.39348,\t y triplet: 21.28632\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 39.53329372406006 seconds\n",
      "Jaccard graph constructed in 33.72168827056885 seconds\n",
      "Wrote graph to binary file in 3.9888854026794434 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.60605\n",
      "Louvain completed 21 runs in 19.598812341690063 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 122.24392032623291 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 13.870426416397095 seconds\n",
      "Jaccard graph constructed in 35.32652926445007 seconds\n",
      "Wrote graph to binary file in 2.292210340499878 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.766059\n",
      "After 8 runs, maximum modularity is Q = 0.767504\n",
      "Louvain completed 28 runs in 35.1239492893219 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 111.53403449058533 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 1058.07507,\t reconstruction loss: 843.50092,\t sparse penalty: 0.28187,\t x triplet loss: 21.34587,\t y triplet loss: 21.28708\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.717998504638672 seconds\n",
      "Jaccard graph constructed in 34.69157862663269 seconds\n",
      "Wrote graph to binary file in 2.79597806930542 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.592042\n",
      "After 3 runs, maximum modularity is Q = 0.596166\n",
      "Louvain completed 23 runs in 23.830042600631714 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 96.3738329410553 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.267264604568481 seconds\n",
      "Jaccard graph constructed in 33.59955430030823 seconds\n",
      "Wrote graph to binary file in 2.1476006507873535 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.778318\n",
      "Louvain completed 21 runs in 21.89419913291931 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 91.85964179039001 seconds\n",
      "epoch: 50, \t total loss: 1056.59814,\t reconstruction loss: 842.78741,\t sparse penalty: 0.27634,\t x triplet loss: 21.26583,\t y triplet loss: 21.21998\n",
      "epoch: 100, \t total loss: 1048.61719,\t reconstruction loss: 835.13519,\t sparse penalty: 0.25556,\t x triplet loss: 21.24282,\t y triplet loss: 21.19803\n",
      "epoch: 150, \t total loss: 1045.27783,\t reconstruction loss: 832.10120,\t sparse penalty: 0.23761,\t x triplet loss: 21.21991,\t y triplet loss: 21.17780\n",
      "epoch: 200, \t total loss: 1043.70947,\t reconstruction loss: 830.78271,\t sparse penalty: 0.22156,\t x triplet loss: 21.20198,\t y triplet loss: 21.16183\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.871376514434814 seconds\n",
      "Jaccard graph constructed in 33.85512161254883 seconds\n",
      "Wrote graph to binary file in 3.0111820697784424 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.596381\n",
      "After 2 runs, maximum modularity is Q = 0.599544\n",
      "After 15 runs, maximum modularity is Q = 0.601707\n",
      "Louvain completed 35 runs in 34.30886173248291 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 107.31580376625061 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 55.89139246940613 seconds\n",
      "Jaccard graph constructed in 35.307594537734985 seconds\n",
      "Wrote graph to binary file in 2.2004947662353516 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.795639\n",
      "Louvain completed 21 runs in 22.90503215789795 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 141.4944713115692 seconds\n",
      "epoch: 250, \t total loss: 1039.46143,\t reconstruction loss: 826.71234,\t sparse penalty: 0.20869,\t x triplet loss: 21.19189,\t y triplet loss: 21.14924\n",
      "epoch: 300, \t total loss: 1037.18970,\t reconstruction loss: 824.59991,\t sparse penalty: 0.19802,\t x triplet loss: 21.18010,\t y triplet loss: 21.13984\n",
      "epoch: 350, \t total loss: 1034.04138,\t reconstruction loss: 821.61066,\t sparse penalty: 0.18957,\t x triplet loss: 21.16623,\t y triplet loss: 21.13034\n",
      "epoch: 400, \t total loss: 1033.46545,\t reconstruction loss: 821.14972,\t sparse penalty: 0.18237,\t x triplet loss: 21.15819,\t y triplet loss: 21.12258\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 11.61109209060669 seconds\n",
      "Jaccard graph constructed in 44.682074546813965 seconds\n",
      "Wrote graph to binary file in 3.3146800994873047 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.582497\n",
      "After 2 runs, maximum modularity is Q = 0.586983\n",
      "After 9 runs, maximum modularity is Q = 0.588165\n",
      "Louvain completed 29 runs in 37.40324354171753 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 127.6392970085144 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.697559595108032 seconds\n",
      "Jaccard graph constructed in 42.15866732597351 seconds\n",
      "Wrote graph to binary file in 2.3860535621643066 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.797089\n",
      "After 4 runs, maximum modularity is Q = 0.798123\n",
      "After 5 runs, maximum modularity is Q = 0.7997\n",
      "After 15 runs, maximum modularity is Q = 0.801448\n",
      "Louvain completed 35 runs in 51.40357828140259 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 137.8395516872406 seconds\n",
      "epoch: 450, \t total loss: 1030.42065,\t reconstruction loss: 818.30267,\t sparse penalty: 0.18261,\t x triplet loss: 21.13407,\t y triplet loss: 21.10693\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1781.79028,\t reconstruction loss: 1479.68066,\t sparse penalty: 60.42191\n",
      "epoch: 50, \t total loss: 914.10730,\t reconstruction loss: 880.36121,\t sparse penalty: 6.74922\n",
      "epoch: 100, \t total loss: 849.43988,\t reconstruction loss: 848.83752,\t sparse penalty: 0.12048\n",
      "epoch: 150, \t total loss: 833.93005,\t reconstruction loss: 833.54028,\t sparse penalty: 0.07796\n",
      "epoch: 0, \t total loss: 1295.37305,\t reconstruction loss: 1048.01794,\t sparse penalty: 0.26195,\t x triplet: 24.87371,\t y triplet: 24.33538\n",
      "epoch: 50, \t total loss: 1091.56079,\t reconstruction loss: 867.66364,\t sparse penalty: 0.39148,\t x triplet: 22.28186,\t y triplet: 22.10608\n",
      "epoch: 100, \t total loss: 1090.46179,\t reconstruction loss: 868.29663,\t sparse penalty: 0.32838,\t x triplet: 22.13008,\t y triplet: 21.97458\n",
      "epoch: 150, \t total loss: 1086.63806,\t reconstruction loss: 865.40106,\t sparse penalty: 0.28548,\t x triplet: 22.03477,\t y triplet: 21.92716\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 11.239810705184937 seconds\n",
      "Jaccard graph constructed in 40.9538996219635 seconds\n",
      "Wrote graph to binary file in 2.936002731323242 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.707895\n",
      "After 3 runs, maximum modularity is Q = 0.709445\n",
      "Louvain completed 23 runs in 24.112680912017822 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 110.09072971343994 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 64.31728076934814 seconds\n",
      "Jaccard graph constructed in 39.2672016620636 seconds\n",
      "Wrote graph to binary file in 2.3741228580474854 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.813894\n",
      "After 2 runs, maximum modularity is Q = 0.815083\n",
      "Louvain completed 22 runs in 28.38765048980713 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 165.29365491867065 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 1083.99890,\t reconstruction loss: 863.17700,\t sparse penalty: 0.25759,\t x triplet loss: 21.97039,\t y triplet loss: 21.93641\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.810517311096191 seconds\n",
      "Jaccard graph constructed in 41.455552101135254 seconds\n",
      "Wrote graph to binary file in 2.6655519008636475 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.707819\n",
      "Louvain completed 21 runs in 20.023538827896118 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 106.23607754707336 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 9.955719947814941 seconds\n",
      "Jaccard graph constructed in 39.67458248138428 seconds\n",
      "Wrote graph to binary file in 2.125744104385376 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.817624\n",
      "After 5 runs, maximum modularity is Q = 0.82096\n",
      "After 15 runs, maximum modularity is Q = 0.822671\n",
      "Louvain completed 35 runs in 42.68135452270508 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 125.85181212425232 seconds\n",
      "epoch: 50, \t total loss: 1081.68982,\t reconstruction loss: 861.54694,\t sparse penalty: 0.24651,\t x triplet loss: 21.90477,\t y triplet loss: 21.87730\n",
      "epoch: 100, \t total loss: 1078.96606,\t reconstruction loss: 859.10315,\t sparse penalty: 0.22691,\t x triplet loss: 21.88544,\t y triplet loss: 21.86024\n",
      "epoch: 150, \t total loss: 1077.17407,\t reconstruction loss: 857.55444,\t sparse penalty: 0.20943,\t x triplet loss: 21.86879,\t y triplet loss: 21.84571\n",
      "epoch: 200, \t total loss: 1074.55457,\t reconstruction loss: 855.09753,\t sparse penalty: 0.19580,\t x triplet loss: 21.85931,\t y triplet loss: 21.83629\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.01651644706726 seconds\n",
      "Jaccard graph constructed in 43.87714147567749 seconds\n",
      "Wrote graph to binary file in 2.4156911373138428 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.698973\n",
      "After 2 runs, maximum modularity is Q = 0.700235\n",
      "After 8 runs, maximum modularity is Q = 0.70141\n",
      "Louvain completed 28 runs in 32.13682150840759 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 119.44212532043457 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 210.62942719459534 seconds\n",
      "Jaccard graph constructed in 42.86331653594971 seconds\n",
      "Wrote graph to binary file in 2.4680533409118652 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.838003\n",
      "After 2 runs, maximum modularity is Q = 0.841239\n",
      "Louvain completed 22 runs in 26.13739013671875 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 314.6241693496704 seconds\n",
      "epoch: 250, \t total loss: 1067.80371,\t reconstruction loss: 848.36743,\t sparse penalty: 0.19094,\t x triplet loss: 21.86470,\t y triplet loss: 21.83162\n",
      "epoch: 300, \t total loss: 1064.80029,\t reconstruction loss: 845.58307,\t sparse penalty: 0.18265,\t x triplet loss: 21.84586,\t y triplet loss: 21.81494\n",
      "epoch: 350, \t total loss: 1062.46252,\t reconstruction loss: 843.40088,\t sparse penalty: 0.17511,\t x triplet loss: 21.83306,\t y triplet loss: 21.80415\n",
      "epoch: 400, \t total loss: 1060.70093,\t reconstruction loss: 841.76233,\t sparse penalty: 0.16762,\t x triplet loss: 21.82388,\t y triplet loss: 21.79622\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 9.897135496139526 seconds\n",
      "Jaccard graph constructed in 42.0938937664032 seconds\n",
      "Wrote graph to binary file in 2.7352633476257324 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.721528\n",
      "After 3 runs, maximum modularity is Q = 0.725064\n",
      "After 14 runs, maximum modularity is Q = 0.726676\n",
      "Louvain completed 34 runs in 38.34710884094238 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 125.50752878189087 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 10.66583514213562 seconds\n",
      "Jaccard graph constructed in 42.479851961135864 seconds\n",
      "Wrote graph to binary file in 2.4644575119018555 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.846173\n",
      "After 2 runs, maximum modularity is Q = 0.848031\n",
      "Louvain completed 22 runs in 24.88404655456543 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 113.69895124435425 seconds\n",
      "epoch: 450, \t total loss: 1058.74329,\t reconstruction loss: 839.89917,\t sparse penalty: 0.16116,\t x triplet loss: 21.81593,\t y triplet loss: 21.79173\n",
      "++++++++++ MUSE completed ++++++++++\n"
     ]
    }
   ],
   "source": [
    "## this is just for testing\n",
    "\n",
    "################################# loop batch version, random sample\n",
    "for i in range(5):\n",
    "\n",
    "  np.random.seed(i)\n",
    "  indices = np.random.choice(datab_full.shape[0], 25000, replace=False)\n",
    "  dataa_sub = dataa.iloc[indices,:]\n",
    "  datab_sub = datab_full[indices,:]\n",
    "  view_a_label_sub = view_a_label[indices]\n",
    "  view_b_label_sub = view_b_label[indices]\n",
    "\n",
    "  muse_feature, reconstruct_x, reconstruct_y, \\\n",
    "    latent_x, latent_y = muse_fit_predict(dataa_sub.to_numpy(),\n",
    "                                          datab_sub,\n",
    "                                          view_a_label_sub,\n",
    "                                          view_b_label_sub,\n",
    "                                          latent_dim=30,\n",
    "                                          n_epochs=500,\n",
    "                                          lambda_regul=5,\n",
    "                                          lambda_super=5)\n",
    "  name = 'batch_' + str(i) + '.npy'\n",
    "  np.save('../data/muse_xton_embedding_' + name, muse_feature)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
